From 4d1ab764a2e897ef191d6dab2de4f1c139fba994 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marek=20Marczykowski-G=C3=B3recki?=
 <marmarek@invisiblethingslab.com>
Date: Tue, 10 Feb 2015 02:20:06 +0100
Subject: [PATCH] libxl: allow libxl_domain_suspend to simply suspend a domain,
 without a savefile
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
Organization: Invisible Things Lab
Cc: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>

When fd=-1, no savefile will be written, but the domain will still be
suspended (but not destroyed). This works pretty well on PV domains, but
is totally untested on HVM (most likely it will crash device model).

Signed-off-by: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.com>

Ported to Xen 4.7.0
Signed-off-by: Marcus of Wetware Labs <marcus@wetwa.re>
---
 tools/libxl/libxl.c       | 29 +++++++++++++++++++++--------
 tools/libxl/xl.h          |  2 ++
 tools/libxl/xl_cmdimpl.c  | 47 +++++++++++++++++++++++++++++++++++++++++++++++
 tools/libxl/xl_cmdtable.c | 10 ++++++++++
 4 files changed, 80 insertions(+), 8 deletions(-)

diff --git a/tools/libxl/libxl.c b/tools/libxl/libxl.c
index d38d0c7..ef3c79f 100644
--- a/tools/libxl/libxl.c
+++ b/tools/libxl/libxl.c
@@ -954,9 +954,11 @@ static void domain_suspend_cb(libxl__egc *egc,
     STATE_AO_GC(dss->ao);
     int flrc;
 
-    flrc = libxl__fd_flags_restore(gc, dss->fd, dss->fdfl);
-    /* If suspend has failed already then report that error not this one. */
-    if (flrc && !rc) rc = flrc;
+    if (dss->fd >= 0) {
+        flrc = libxl__fd_flags_restore(gc, dss->fd, dss->fdfl);
+        /* If suspend has failed already then report that error not this one. */
+        if (flrc && !rc) rc = flrc;
+    }
 
     libxl__ao_complete(egc,ao,rc);
 
--- xen-4.7.0/tools/libxl/libxl.c.orig	2016-07-17 20:24:05.484000000 +0300
+++ xen-4.7.0/tools/libxl/libxl.c	2016-07-17 21:01:58.717000000 +0300
@@ -934,6 +934,13 @@
 
 }
 
+static void domain_suspend_empty_cb(libxl__egc *egc,
+                              libxl__domain_suspend_state *dss, int rc)
+{
+    STATE_AO_GC(dss->ao);
+    libxl__ao_complete(egc,ao,rc);
+}
+
 int libxl_domain_suspend(libxl_ctx *ctx, uint32_t domid, int fd, int flags,
                          const libxl_asyncop_how *ao_how)
 {
@@ -946,25 +953,40 @@
         goto out_err;
     }
 
-    libxl__domain_save_state *dss;
-    GCNEW(dss);
-
-    dss->ao = ao;
-    dss->callback = domain_suspend_cb;
-
-    dss->domid = domid;
-    dss->fd = fd;
-    dss->type = type;
-    dss->live = flags & LIBXL_SUSPEND_LIVE;
-    dss->debug = flags & LIBXL_SUSPEND_DEBUG;
-    dss->checkpointed_stream = LIBXL_CHECKPOINTED_STREAM_NONE;
-
-    rc = libxl__fd_flags_modify_save(gc, dss->fd,
-                                     ~(O_NONBLOCK|O_NDELAY), 0,
-                                     &dss->fdfl);
-    if (rc < 0) goto out_err;
+    if (fd >= 0) {
+        libxl__domain_save_state *dss;
+        GCNEW(dss);
+
+        dss->ao = ao;
+        dss->callback = domain_suspend_cb;
+
+        dss->domid = domid;
+        dss->fd = fd;
+        dss->type = type;
+        dss->live = flags & LIBXL_SUSPEND_LIVE;
+        dss->debug = flags & LIBXL_SUSPEND_DEBUG;
+        dss->checkpointed_stream = LIBXL_CHECKPOINTED_STREAM_NONE;
+
+        rc = libxl__fd_flags_modify_save(gc, dss->fd,
+                                         ~(O_NONBLOCK|O_NDELAY), 0,
+                                         &dss->fdfl);
+        if (rc < 0) goto out_err;
+
+        libxl__domain_save(egc, dss);
+    } else {
+        libxl__domain_suspend_state *dsps;
+        GCNEW(dsps);
+        dsps->ao = ao;
+        dsps->domid = domid;
+        dsps->type = type;
+        dsps->guest_evtchn.port = -1;
+        dsps->guest_evtchn_lockfd = -1;
+        dsps->guest_responded = 0;
+        /* TODO: setup dss->guest_evtchn.port ? */
+        dsps->callback_common_done = domain_suspend_empty_cb;
+        libxl__domain_suspend(egc, dsps);
+    }
 
-    libxl__domain_save(egc, dss);
     return AO_INPROGRESS;
 
  out_err:
diff --git a/tools/libxl/xl.h b/tools/libxl/xl.h
index 13bccba..8f36a44 100644
--- a/tools/libxl/xl.h
+++ b/tools/libxl/xl.h
@@ -48,6 +48,8 @@ int main_restore(int argc, char **argv);
 int main_migrate_receive(int argc, char **argv);
 int main_save(int argc, char **argv);
 int main_migrate(int argc, char **argv);
+int main_suspend(int argc, char **argv);
+int main_resume(int argc, char **argv);
 #endif
 int main_dump_core(int argc, char **argv);
 int main_pause(int argc, char **argv);
diff --git a/tools/libxl/xl_cmdimpl.c b/tools/libxl/xl_cmdimpl.c
index 9a2870e..83e9a5b 100644
--- a/tools/libxl/xl_cmdimpl.c
+++ b/tools/libxl/xl_cmdimpl.c
@@ -4096,6 +4096,27 @@ static int save_domain(uint32_t domid, const char *filename, int checkpoint,
     exit(rc < 0 ? EXIT_FAILURE : EXIT_SUCCESS);
 }
 
+static int suspend_domain(uint32_t domid)
+{
+    int rc = libxl_domain_suspend(ctx, domid, -1, 0, NULL);
+
+    if (rc < 0) {
+        fprintf(stderr, "Failed to suspend domain, resuming domain\n");
+        libxl_domain_resume(ctx, domid, 1, 0);
+    }
+    exit(rc < 0 ? 1 : 0);
+}
+
+static int resume_domain(uint32_t domid)
+{
+    int rc = libxl_domain_resume(ctx, domid, 1, 0);
+
+    if (rc < 0) {
+        fprintf(stderr, "Failed to resume domain\n");
+    }
+    exit(rc < 0 ? 1 : 0);
+}
+
 static pid_t create_migration_child(const char *rune, int *send_fd,
                                         int *recv_fd)
 {
@@ -4849,6 +4870,32 @@ int main_reboot(int argc, char **argv)
     return main_shutdown_or_reboot(1, argc, argv);
 }
 
+int main_suspend(int argc, char **argv)
+{
+    int opt;
+
+    SWITCH_FOREACH_OPT(opt, "", NULL, "suspend", 1) {
+        /* No options */
+    }
+
+    suspend_domain(find_domain(argv[optind]));
+
+    return 0;
+}
+
+int main_resume(int argc, char **argv)
+{
+    int opt;
+
+    SWITCH_FOREACH_OPT(opt, "", NULL, "resume", 1) {
+        /* No options */
+    }
+
+    resume_domain(find_domain(argv[optind]));
+
+    return 0;
+}
+
 int main_list(int argc, char **argv)
 {
     int opt;
diff --git a/tools/libxl/xl_cmdtable.c b/tools/libxl/xl_cmdtable.c
index 0071f12..12e28eb 100644
--- a/tools/libxl/xl_cmdtable.c
+++ b/tools/libxl/xl_cmdtable.c
@@ -75,6 +75,16 @@ struct cmd_spec cmd_table[] = {
       "                        no PV drivers.\n"
       "-w, --wait              Wait for guest(s) to shutdown.\n"
     },
+    { "suspend",
+      &main_suspend, 0, 1,
+      "Suspend a domain",
+      "[options] <Domain>",
+    },
+    { "resume",
+      &main_resume, 0, 1,
+      "Resume suspended domain",
+      "[options] <Domain>",
+    },
     { "reboot",
       &main_reboot, 0, 1,
       "Issue a reboot signal to a domain",
-- 
2.1.0

